### PR TITLE
Add dataclass validators

### PR DESCRIPTION
**Pull request by samet-akcay**
## 📝 Description

- Add validators

> TODO:
- [ ] Inherit `Validator` as a mixin class for each dataclass to not implement the validate methods explicitly.

> WARNING
When reviewing the PR, it would be great to be as critical as possible, especially for the validator implementations.

## ✨ Changes

Select what type of change your PR is:

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔨 Refactor (non-breaking change which refactors the code base)
- [ ] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🔒 Security update

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [ ] 📋 I have summarized my changes in the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) and followed the guidelines for my type of change (skip for minor changes, documentation updates, and test enhancements).
- [ ] 📚 I have made the necessary updates to the documentation (if applicable).
- [ ] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).

For more information about code review checklists, see the [Code Review Checklist](https://github.com/openvinotoolkit/anomalib/blob/main/docs/source/markdown/guides/developer/code_review_checklist.md).

### CODE REVIEW

File: src/anomalib/data/validators/numpy/__init__.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+from .depth import NumpyDepthBatchValidator, NumpyDepthValidator
+from .image import NumpyImageBatchValidator, NumpyImageValidator
+from .video import NumpyVideoBatchValidator, NumpyVideoValidator
+
+__all__ = [
+    "NumpyImageBatchValidator",
+    "NumpyImageValidator",
```
Comment (samet-akcay) [inline src/anomalib/data/validators/numpy/__init__.py:12]: @ashwinvaidya17, @djdameln, which one would you prefer?
`"NumpyImageValidator"` or `"NumpyImageItemValidator"`

File: tests/unit/data/validators/torch/test_image.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        with pytest.raises(ValueError, match="Ground truth label must be a scalar"):
+            self.validator.validate_gt_label(torch.tensor([0, 1]))
+
+    def test_validate_gt_mask_valid(self) -> None:
+        """Test validation of a valid ground truth mask."""
+        mask = torch.randint(0, 2, (1, 224, 224))
+        validated_mask = self.validator.validate_gt_mask(mask)
+        assert isinstance(validated_mask, Mask)
+        assert validated_mask.shape == (224, 224)
+        assert validated_mask.dtype == torch.bool
```
Comment (ashwinvaidya17) [inline tests/unit/data/validators/torch/test_image.py:72]: Same. We convert int to bool. Not sure if the users will realise that they have to use the inputs returned from the validate method.

**Review by ashwinvaidya17 — COMMENTED** [review]
Thanks for the massive effort. I have a few general comments.
In the review I've ignored the numpy validators as I feel we can remove those.
1. There is some duplication between Image, Video, and Depth validators. Is there a way we can reduce those? It might be harder to maintain in the future.
2. Some of the validators change the inputs passed to them. Maybe we can use a different term than `validate`? How about `sanitize`?

**Comment by samet-akcay** [comment]
Merging this as is. Will address the comments in subsequent PRs

**Pull request merged**